src/runtime_src/core/common/api/aie/xrt_graph.cpp

```diff
@@ -112,6 +112,59 @@ class graph_impl
 
 }
 
+namespace xrt { namespace aie {
+
+class profiling_impl
+{
+private:
+  std::shared_ptr<xrt_core::device> device;
+  int profiling_hdl;
+
+public:
+  using handle = int;
+  static constexpr handle invalid_handle = -1;
+
+  profiling_impl(std::shared_ptr<xrt_core::device> dev)
+    : device(std::move(dev)),
+      profiling_hdl(invalid_handle)
+  {}
+
+  ~profiling_impl()
+  {
+    device->stop_profiling(profiling_hdl);
+  }
```

Comment:
Uh, in conjunction with user callling `stop()` this dtor won't work.  I think you need a bit more logic to the state so stop_profiling can bail if it was already called.  My point earlier was that you must handle the case where user forgets to call `stop()` but you must also handle the case where user does call `stop()` and the handle is invalid.

Destructors cannot throw, so you must have a  try/catch if the function that is called can throw.